### DISCUSSION TITLE
Turn off SpanCharacteristic from code?

### DISCUSSION THREAD

**Discussion by chrwh** [HomeSpan/HomeSpan#339] [Q&A]
Hi,

Thanks for a SUPER lib, it's working so great :-)

How can I turn off a SpanCharacteristic (Characteristic::Name("All Off");) from code ?
I need this because when Characteristic "a" is turned on from the Apple device Characteristic "b" should be turned off, I know I could make an automation on the device, but I would rather do it in the kode.

Any help would be a great help.

Best regards,
Christian

**Comment by HomeSpan** [comment]
You can use the setVal() method to change the value of any Characteristic automatically in the code.  This can be done, for example, in response to a button press, in response to a periodic timer, or in response to another Characteristic being changed via the Home App. 

For instance, you may have a master switch that shuts off other switches, or a Fan that turns off a light, etc.  

Simply use setVal in the update() portion of a Service to program whatever actions you would like based on whatever updates come from the Home App.

**Reply by chrwh** [reply]
Thanks for the answer, but i'm in doubt, how I specify the correct one to turn on/off?

**Reply by HomeSpan** [reply]
Can you provide some details on exactly what you are trying to do?  One of the tutorial examples may be helpful.

**Reply by chrwh** [reply]
I have 14 different Characteristic when a certin Characteristic is turned on from homekit device, I need another Characteristic to turn off from inside my code

**Reply by HomeSpan** [reply]
What Service are you using that has 14 Characteristics?

**Reply by chrwh** [reply]
Well I have only worked with this for a single night, so I might have misunderstod something.
What I want is 1 main tile, with 14 sub tiles with switches that are momentary, except for a few of them. But I also need to control some of the switches from inside the code. So when CD is on, and AUX i set to on then CD needs to go to off, and vice versa. I'm making a remote (I have seen the remote example, but that is not quite what I wan't). Later i need to add RMT at 455KHz, but I might ask about that later.
I have attached the code
[BeoSound9000Control.zip](https://github.com/HomeSpan/HomeSpan/files/9076107/BeoSound9000Control.zip)

**Reply by chrwh** [reply]
So looking more at this, it seems, that I have 14 switches, so now I think the question is, how to manipulate a switch from the code, then again I could be wrong again ;-)

Service                             UUID         AID  IID  Update  Loop  Button  Linked Services
------------------------------  --------  ----------  ---  ------  ----  ------  ---------------
AccessoryInformation                  3E           1    1      NO    NO      NO  -
Switch                                49           1    7     YES   YES      NO  -
Switch                                49           1   10     YES   YES      NO  -
Switch                                49           1   13     YES   YES      NO  -
Switch                                49           1   16     YES   YES      NO  -
Switch                                49           1   19     YES   YES      NO  -
Switch                                49           1   22     YES   YES      NO  -
Switch                                49           1   25     YES   YES      NO  -
Switch                                49           1   28     YES   YES      NO  -
Switch                                49           1   31     YES   YES      NO  -
Switch                                49           1   34     YES   YES      NO  -
Switch                                49           1   37     YES   YES      NO  -
Switch                                49           1   40     YES   YES      NO  -
Switch                                49           1   43     YES   YES      NO  -
Switch                                49           1   46     YES   YES      NO  -

**Reply by HomeSpan** [reply]
Yes, it appears that you created 14 instances of a Service::Switch (section 8.38 of the HAP-R2 document).  Each instance should have a Characteristic::On specifying whether the switch is on or off (section 9.70 of HAP-R2).  If you go through all the different tutorial examples you'll see that when creating either a Service or a Characteristic you can do so without saving it, such as `new Service:: MySwitch()` or by saving it in a pointer, such as `SpanService *myService = new Service::MySwitch()`.  You can also simply instantiate  something without using a pointer, such as `Characteristic::On powerStatus()`. You will probably want to create an array of Service pointers to store all 14 of your Services.  There are many ways to do this in C++ from creating a simple array to using a vector, etc.

Then, in the update() portion of one of your "master" service (the one that will trigger other services), you can say something like: `myService[4]->powerStatus.setVal(0)` to turn off the 4th switch.  Whether you use the `.` or the `->` will depend on exactly how you structure your variables as either pointers or not.

This is just a high level summary - going through the tutorials will show you how this works in practice using a wide variety of examples.  

As an aside, HomeKit is not actually well-suited for duplicating a traditional many-button remote control, such as operating a TV.  It's fine for more simple multi-button remotes, with just on/off switches, but duplicating channel numbers, volume up/down, etc. requires a lot of clunky workarounds.  The main problem is that HomeKit does not have native functionality to duplicate a simple push button in the Home App.  Everything is based on switches instead.

**Reply by chrwh** [reply]
Thank you so much for the answer. Unfortunately it still not makes sense to me how to make those 14 pointers.
So I might just have to go another way
Best regards,
Christian

**Reply by chrwh** [reply]
I have now been thru all the examples, but I don't see an example on how to put all the switches into a array, so I can get "hold" of them later. Can you provide an example on how to do this? I can't simply wrap my head arround that one :-(

**Reply by chrwh** [reply]
Help still needed. I have tried a lot of different things, but have still not had any luck. 
anyone has a suggestion?
